packages/hooks/src/createRemoveHook.js

```diff
@@ -0,0 +1,44 @@
+/**
+ * Returns a function which, when invoked, will remove a specified hook or all
+ * hooks by the given name.
+ *
+ * @param  {Object}   hooks      Stored hooks, keyed by hook name.
+ * @param  {bool}     removeAll  Whether to remove all hooked callbacks.
+ *
+ * @return {Function}            Function that removes hooks.
+ */
+function createRemoveHook( hooks, removeAll ) {
+	/**
+	 * Removes the specified callback (or all callbacks) from the hook with a
+	 * given name.
+	 *
+	 * @param {string}    hookName The name of the hook to modify.
+	 * @param {?Function} callback The specific callback to be removed.  If
+	 *                             omitted (and `removeAll` is truthy), clears
+	 *                             all callbacks.
+	 */
+	return function removeHook( hookName, callback ) {
+		// Bail if no hooks exist by this name
+		if ( ! hooks.hasOwnProperty( hookName ) ) {
+			return;
+		}
+
+		if ( removeAll ) {
+			const runs = hooks[ hookName ].runs;
+			hooks[ hookName ] = [];
+			if ( runs ) {
+				hooks[ hookName ].runs = runs;
```

Comment:
I think this approach will be cleaner as well.  However, I think we have to avoid `for ... in` because we can't guarantee that plugins aren't extending object prototypes.